tableauserverclient/server/endpoint/datasources_endpoint.py

```diff
@@ -252,11 +283,13 @@ def publish(
         else:
             logger.info("Publishing {0} to server".format(filename))
 
-            try:
+            if isinstance(file, (Path, str)):
                 with open(file, "rb") as f:
                     file_contents = f.read()
-            except TypeError:
+            elif isinstance(file, io_types):
                 file_contents = file.read()
+            else:
+                raise TypeError()
```

Comment:
Should we have some sort of string in the exception like "the thingy isn't a file or a path"